api/scheduler/environment_scaler.go

```diff
@@ -0,0 +1,130 @@
+package scheduler
+
+import (
+	"fmt"
+	"github.com/quintilesims/layer0/api/scheduler/resource"
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+)
+
+type EnvironmentScaler interface {
+	Scale(environmentID string) (*models.ScalerRunInfo, error)
+}
+
+type L0EnvironmentScaler struct {
+	consumerGetter  resource.ConsumerGetter
+	providerManager resource.ProviderManager
+}
+
+func NewL0EnvironmentScaler(c resource.ConsumerGetter, p resource.ProviderManager) *L0EnvironmentScaler {
+	return &L0EnvironmentScaler{
+		consumerGetter:  c,
+		providerManager: p,
+	}
+}
+
+func (r *L0EnvironmentScaler) Scale(environmentID string) (*models.ScalerRunInfo, error) {
+	resourceProviders, err := r.providerManager.GetProviders(environmentID)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceConsumers, err := r.consumerGetter.GetConsumers(environmentID)
+	if err != nil {
+		return nil, err
+	}
+
+	// todo: get run info
+	return RunBasicScaler(environmentID, resourceProviders, resourceConsumers, r.providerManager)
+}
+
+func RunBasicScaler(
+	environmentID string,
+	providers []*resource.ResourceProvider,
+	consumers []resource.ResourceConsumer,
+	providerManager resource.ProviderManager,
+) (*models.ScalerRunInfo, error) {
+
+	scaleBeforeRun := len(providers)
+	var errs []error
+
+	// check if we need to scale up
+	for _, consumer := range consumers {
+		hasRoom := false
+
+		// first, sort by memory so we pack tasks by memory as tightly as possible
+		resource.SortProvidersByMemory(providers)
+
+		// next, place any unused providers in the back of the list
+		// that way, we can can delete them if we avoid placing any tasks in them
+		resource.SortProvidersByUsage(providers)
+
+		for _, provider := range providers {
+			if provider.HasResourcesFor(consumer) {
+				hasRoom = true
+				provider.SubtractResourcesFor(consumer)
+				break
+			}
+		}
+
+		if !hasRoom {
+			newProvider, err := providerManager.AddNewProvider(environmentID)
+			if err != nil {
+				return nil, err
+			}
+
+			if !newProvider.HasResourcesFor(consumer) {
+				err := fmt.Errorf("Resource '%s' cannot fit into an empty provider!", consumer.ID)
```

Comment:
let's change this error message to be something more meaningful for the end-user... i.e. This Environment Instance Type is Too Small....